### DISCUSSION TITLE
unable to add colons in tags

### DISCUSSION THREAD

**Discussion by mbalandis** [downforcetech/fluidxml.php#8] [Questions]
I cannot uderstand how to add colons : in the tags. It works fine in attributes. See code example below. Am I missing something?  Part that doesn't work: add('atom:link', '...')

```
        use \FluidXml\FluidXml;
        use \FluidXml\FluidNamespace;

        $feed = new FluidXml(null, ['root' => 'rss']);
              $feed->setAttribute(['version' => '2.0', 'xmlns:atom' => 'http://www.w3.org/2005/Atom',                   'xmlns:itunes' => 'http://www.itunes.com/dtds/podcast-1.0.dtd']);     

         $feed->addChild('channel', true)
                   ->add('atom:link', '...')
                   ->addChild('title', 'Title');

        echo $feed;
```

error i am getting: Fatal error: Call to a member function uri() on null

**Comment by mbalandis** [comment]
Ok, so I figured that you need to use namespaces for that, however, it seems to have an issue not letting name 2 of the same.

For example: 

`
$feed->namespace('xmlns', 'http://www.w3.org/2005/Atom',         FluidNamespace::MODE_IMPLICIT)->namespace('xmlns', 'http://www.itunes.com/dtds/podcast-1.0.dtd', FluidNamespace::MODE_IMPLICIT);`

and my xml will endup with either _http://www.w3.org/2005/Atom_ or _http://www.itunes.com/dtds/podcast-1.0.dtd_, latter will overwrite it. 

and to get colon work i need to write something like this: xmlns:atom:link and xmlns:itunes:image
it works but the attribute xmlns:itunes or xmlns:atom in the tags are not needed, let alone are incorrect.

Example output (i have some extra stuff added):

`<?xml version="1.0" encoding="UTF-8"?>
<rss version="2.0" xmlns:atom="http://www.w3.org/2005/Atom" xmlns:itunes="http://www.itunes.com/dtds/podcast-1.0.dtd">
  <channel>
    <atom:link xmlns:atom="http://www.itunes.com/dtds/podcast-1.0.dtd" href="http://thisislink.com" type="application/rss+xml" rel="self"/>
    <itunes:image xmlns:itunes="http://www.itunes.com/dtds/podcast-1.0.dtd" href="1234.png"/>
    <image>
      <title>FEED TITLE</title>
      <url>1234.png</url>
    </image>
    <title>aaaaaaa</title>
  </channel>
</rss>
`

it adds: xmlns:atom="http://www.itunes.com/dtds/podcast-1.0.dtd" and xmlns:itunes="http://www.itunes.com/dtds/podcast-1.0.dtd" which are not needed and wrong as I can set 1 per xmlns by defining namespaces.

Hope that makes sense. Other than that the library is almost perfect.

**Comment by daniele-orlando** [comment]
About the first comment, as you guessed it is not possible in xml to use semi colon in tag name because it is reserved for the namespace part.

About the second comment, please check out the wiki here
https://github.com/servo-php/fluidxml/wiki/Getting-Started:-8.-Namespaces
at the paragraph starting with

> That's it! Even XML namespaces can be easy and fun to use....

It solves exactly what you need, just follow the documentation example.

In the next release there will be a constructor option that will allow to register namespaces with the constructor, helping in situations like your. For now, just use the code explained above and stay tuned.
